src/custom/state/orders/updaters/EventUpdater.ts

```diff
@@ -0,0 +1,85 @@
+import { useCallback, useEffect, useRef } from 'react'
+
+import { useActiveWeb3React } from 'hooks/web3'
+
+import { useCancelOrdersBatch, useExpireOrdersBatch, useFulfillOrdersBatch, usePendingOrders } from 'state/orders/hooks'
+import { ApiOrderStatus } from 'state/orders/utils'
+import { OrderFulfillmentData } from 'state/orders/actions'
+import { OPERATOR_API_POLL_INTERVAL } from 'state/orders/consts'
+
+import { SupportedChainId as ChainId } from 'constants/chains'
+
+import { OrderID } from 'utils/operator'
+
+import { fetchOrderPopupData, OrderLogPopupMixData } from 'state/orders/updaters/utils'
+
+export function EventUpdater(): null {
+  const { chainId } = useActiveWeb3React()
+
+  const pending = usePendingOrders({ chainId })
+
+  // Ref, so we don't rerun useEffect
+  const pendingRef = useRef(pending)
+  pendingRef.current = pending
+
+  const fulfillOrdersBatch = useFulfillOrdersBatch()
+  const expireOrdersBatch = useExpireOrdersBatch()
+  const cancelOrdersBatch = useCancelOrdersBatch()
+
+  const updateOrders = useCallback(
+    async (chainId: ChainId) => {
+      // Exit early when there are no pending orders
+      if (pendingRef.current.length === 0) {
+        return
+      }
+
+      // Iterate over pending orders fetching operator order data, async
+      const unfilteredOrdersData = await Promise.all(
+        pendingRef.current.map(async (orderFromStore) => fetchOrderPopupData(orderFromStore, chainId))
+      )
+
+      // Group resolved promises by status
+      // Only pick the status that are final
+      const { fulfilled, expired, cancelled } = unfilteredOrdersData.reduce<
+        Record<ApiOrderStatus, OrderLogPopupMixData[]>
+      >(
+        (acc, { status, popupData }) => {
+          popupData && acc[status].push(popupData)
+          return acc
+        },
+        { fulfilled: [], expired: [], cancelled: [], unknown: [], pending: [] }
+      )
+
+      // Bach state update per group, if any
+
+      fulfilled.length > 0 &&
```

Comment:
this wont clash with the CancelledOrdersUpdater logic touching similar state?

Reply:
Well, not clash, but could have some race conditions, yes...

The (not very aptly named `EventUpdater`) checks for pending orders, and update their statuses, triggering toast notifications when they change to fulfilled, expired or cancelled.

The CancelledOrdersUpdater checks only for cancelled orders that might have been set to cancelled too soon.
It can happen that an order is considered `cancelled` but a solver submit a solution to it anyway.
And only for 5min after an order is cancelled.

The last (and new, introduced here) is the UnfillableOrdersUpdater, that checks the price of pending orders and updates a flag in the storage.